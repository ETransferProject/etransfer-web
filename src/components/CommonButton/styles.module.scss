@import '/src/assets/theme/color.scss';
@import '/src/assets/theme/font.scss';

.common-button {
  border-radius: 6px;
  box-shadow: none;
  span {
    @extend .font-family-medium;
  }
}

.primary {
  background: $brand;
  border: 1px solid $brand;
  color: $white-text;
  &:hover {
    background: $hover;
    border-color: $hover;
    color: $white-text;
  }
  &:checked {
    background: $click;
    border-color: $click;
    color: $white-text;
  }
  &:disabled,
  &[disabled]:hover {
    background: $disable;
    border-color: $disable;
    color: $white-text;
  }
  &:focus {
    background: $click;
    border-color: $click;
    color: $white-text;
  }
}

.secondary {
  background: $normal-bg;
  border: 1px solid $normal-bg;
  color: $primary-text;
  &:hover {
    background: $hover-bg;
    border-color: $hover-bg;
    color: $primary-text;
  }
  &:checked {
    background: $click-bg;
    border-color: $click-bg;
    color: $primary-text;
  }
  &:disabled {
    background: $hover-bg;
    border-color: $hover-bg;
    color: $primary-text;
  }
  &:focus {
    background: $click-bg;
    border-color: $click-bg;
    color: $primary-text;
  }
}

.border {
  background: $page-bg !important;
  border: 1px solid $border !important;
  color: $primary-text !important;
  display: flex;
  align-items: center;
  &:hover {
    background: $page-bg !important;
    border-color: $hover !important;
    color: $primary-text !important;
  }
  &:checked {
    background: $page-bg;
    border-color: $click;
    color: $primary-text;
  }
  &:disabled {
    background: $disable;
    border-color: $disable;
    color: $primary-text;
  }
  &:focus {
    background: $page-bg;
    border-color: $click;
    color: $primary-text;
  }
}

.extraSmall {
  height: 32px;
  padding: 0 16px;
  span {
    display: inline-flex;
    font-size: 12px;
    line-height: 20px;
  }
}

.small {
  height: 40px;
  padding: 0 20px;
  span {
    display: inline-flex;
    font-size: 14px;
    line-height: 22px;
  }
}

.middle {
  height: 48px;
  padding: 0 28px;
  span {
    display: inline-flex;
    font-size: 16px;
    line-height: 24px;
  }
}

.large {
  height: 56px;
  padding: 0 32px;
  span {
    display: inline-flex;
    font-size: 16px;
    line-height: 24px;
  }
}

.stretched {
  width: 100%;
}
