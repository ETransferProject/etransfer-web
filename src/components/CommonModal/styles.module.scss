@import 'src/assets/theme/color.scss';
@import 'src/assets/theme/font.scss';

.common-modal {
  :global {
    .ant-modal-content {
      border-radius: 8px;
    }
    .ant-modal-close-x {
      display: flex;
      justify-content: center;
      align-items: center;
      width: calc(20px + (16px * 2));

      svg path {
        fill: $primary-text;
      }
    }
    .ant-modal-header {
      height: 54px;
      border-radius: 8px 8px 0 0;
      border: none;
      text-align: center;
      .ant-modal-title {
        color: $primary-text;
        font-size: 16px;
        line-height: 24px;
        font-weight: $font-family-medium-weight;
        font-family: $font-family-medium-family;
      }
    }
    .ant-modal-body {
      padding: 0 24px 24px;
    }
  }
}

.footer {
  gap: 12px;
  width: 100%;
  margin-top: 32px;
}
