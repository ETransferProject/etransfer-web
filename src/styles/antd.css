html {
  --ant-primary-color: #285eff;
  --ant-primary-color-hover: #3b6cff;
  --ant-primary-color-active: #1e52f0;
  --ant-success-color: #00b73e;
  --ant-error-color: #f53f3f;
  --ant-error-color-hover: #fc7373;
  --ant-warning-color-active: #cf3737;
  --ant-warning-color: #f90;
  --ant-info-color: #285eff;
}
