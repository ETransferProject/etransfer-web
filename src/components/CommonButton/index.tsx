import { Button, ButtonProps } from 'antd';
import clsx from 'clsx';
import styles from './styles.module.scss';

export enum CommonButtonSize {
  ExtraSmall = 'extraSmall',
  Small = 'small',
  Middle = 'middle',
  Large = 'large',
}

export enum CommonButtonType {
  Primary = 'primary',
  Secondary = 'secondary',
}

export type CommonButtonProps = Omit<ButtonProps, 'size' | 'type'> & {
  size?: CommonButtonSize;
  type?: CommonButtonType;
  stretched?: boolean;
};

export default function CommonButton({
  size = CommonButtonSize.Middle,
  type = CommonButtonType.Primary,
  stretched = false,
  ...props
}: CommonButtonProps) {
  return (
    <Button
      {...props}
      className={clsx(
        styles['common-button'],
        styles[size],
        styles[type],
        { [styles['stretched']]: stretched },
        props.className,
      )}
    />
  );
}
