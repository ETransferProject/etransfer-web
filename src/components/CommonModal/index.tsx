import clsx from 'clsx';
import { Modal, ModalProps } from 'antd';
import CloseIcon from 'assets/images/close.svg';
import CommonButton, { CommonButtonType } from 'components/CommonButton';
import styles from './styles.module.scss';

export type CommonModalProps = Omit<
  ModalProps,
  'footer' | 'confirmLoading' | 'okButtonProps' | 'okType'
> & {
  footerClassName?: string;
  cancelText?: string;
  okText?: string;
  hideCancelButton?: boolean;
  hideOkButton?: boolean;
  isOkButtonDisabled?: boolean;
  footerSlot?: React.ReactNode;
};

export default function CommonModal({
  className,
  footerClassName,
  cancelText,
  okText,
  hideCancelButton,
  hideOkButton,
  isOkButtonDisabled,
  ...props
}: CommonModalProps) {
  return (
    <Modal
      width={props.width || 480}
      centered
      zIndex={299}
      closeIcon={<CloseIcon />}
      {...props}
      className={clsx(styles['common-modal'], className)}
      // To keep the title height by default
      title={props.title || ' '}
      footer={null}>
      {props.children}
      {(!hideCancelButton || !hideOkButton) && (
        <div className={clsx('flex-row-center', styles['footer'], footerClassName)}>
          {!hideCancelButton && (
            <CommonButton
              className={clsx('flex-1', styles['cancel-button'])}
              type={CommonButtonType.Secondary}
              onClick={props.onCancel}>
              {cancelText || 'Cancel'}
            </CommonButton>
          )}
          {!hideOkButton && (
            <CommonButton
              className={clsx('flex-1', styles['ok-button'])}
              disabled={isOkButtonDisabled}
              onClick={props.onOk}>
              {okText || 'Confirm'}
            </CommonButton>
          )}
        </div>
      )}
      {props.footerSlot}
    </Modal>
  );
}
