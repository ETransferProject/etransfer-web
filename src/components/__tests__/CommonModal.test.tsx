import { render, screen } from '@testing-library/react';
import { vi, describe, expect, test, beforeEach } from 'vitest';
import CommonModal from '../CommonModal';
import { Modal } from 'antd';
import CommonButton from 'components/CommonButton';
import clsx from 'clsx';

// Mock dependencies
vi.mock('antd', async (importOriginal) => {
  const actual = await importOriginal<typeof import('antd')>();
  return {
    ...actual,
    Modal: vi.fn(({ children, ...props }) => (
      <div data-testid="mock-modal" {...props}>
        {children}
      </div>
    )),
  };
});

vi.mock('components/CommonButton', async (importOriginal) => {
  const actual = await importOriginal<typeof import('components/CommonButton')>();
  return {
    ...actual,
    default: vi.fn((props) => <button {...props} data-testid="common-button" />),
  };
});

vi.mock('clsx', () => ({
  default: vi.fn(() => 'mocked-class'),
}));

describe('CommonModal Component', () => {
  const defaultProps = {
    open: true,
    onCancel: vi.fn(),
    onOk: vi.fn(),
    children: <div data-testid="modal-content" />,
  };

  beforeEach(() => {
    vi.clearAllMocks();
  });

  test('renders with default props', () => {
    render(<CommonModal {...defaultProps} />);

    // Check if the Modal component is rendered
    expect(screen.getByTestId('mock-modal')).toBeInTheDocument();
    expect(Modal).toHaveBeenCalledWith(
      expect.objectContaining({
        width: 480,
        centered: true,
        zIndex: 299,
        title: ' ',
      }),
      expect.anything(),
    );
  });

  test('renders custom footer buttons', () => {
    render(<CommonModal {...defaultProps} />);

    // Check if the two buttons are rendered
    const buttons = screen.getAllByTestId('common-button');
    expect(buttons.length).toBe(2);
    expect(buttons[0]).toHaveTextContent('Cancel');
    expect(buttons[1]).toHaveTextContent('Confirm');
  });

  test('hides footer buttons conditionally', () => {
    const { rerender } = render(<CommonModal {...defaultProps} hideCancelButton hideOkButton />);
    // Assert that the buttons are not rendered
    expect(screen.queryByTestId('common-button')).not.toBeInTheDocument();

    // Rerender with hideCancelButton=true
    rerender(<CommonModal {...defaultProps} hideCancelButton />);
    // Assert that the cancel button is not rendered
    expect(screen.getAllByTestId('common-button').length).toBe(1);
  });

  test('handles custom button texts', () => {
    render(<CommonModal {...defaultProps} cancelText="Abort" okText="Proceed" />);

    // Check if the button texts are correct
    const buttons = screen.getAllByTestId('common-button');
    expect(buttons[0]).toHaveTextContent('Abort');
    expect(buttons[1]).toHaveTextContent('Proceed');
  });

  test('disables ok button', () => {
    render(<CommonModal {...defaultProps} isOkButtonDisabled />);

    // Check if the ok button is disabled
    screen.getAllByTestId('common-button')[1];
    expect(CommonButton).toHaveBeenCalledWith(
      expect.objectContaining({ disabled: true }),
      expect.anything(),
    );
  });

  test('merges class names correctly', () => {
    render(
      <CommonModal {...defaultProps} className="custom-modal" footerClassName="custom-footer" />,
    );

    // Check if the class names are merged correctly
    expect(clsx).toHaveBeenCalledWith(expect.stringContaining('common-modal'), 'custom-modal');
    expect(clsx).toHaveBeenCalledWith(
      'flex-row-center',
      expect.stringContaining('footer'),
      'custom-footer',
    );
  });

  test('renders footer slot', () => {
    render(<CommonModal {...defaultProps} footerSlot={<div data-testid="custom-footer" />} />);

    // Check if the custom footer slot is rendered
    expect(screen.getByTestId('custom-footer')).toBeInTheDocument();
  });

  test('handles custom width', () => {
    render(<CommonModal {...defaultProps} width={600} />);

    // Check if the custom width is passed correctly
    expect(Modal).toHaveBeenCalledWith(expect.objectContaining({ width: 600 }), expect.anything());
  });

  test('matches snapshot', () => {
    const { asFragment } = render(<CommonModal {...defaultProps} />);

    // Check if the snapshot matches the rendered component
    expect(asFragment()).toMatchSnapshot();
  });
});
