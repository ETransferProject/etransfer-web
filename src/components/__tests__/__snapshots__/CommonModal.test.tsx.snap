// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`CommonModal Component > matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="mocked-class"
    closeicon="[object Object]"
    data-testid="mock-modal"
    open=""
    title=" "
    width="480"
    zindex="299"
  >
    <div
      data-testid="modal-content"
    />
    <div
      class="mocked-class"
    >
      <button
        class="mocked-class"
        data-testid="common-button"
        type="secondary"
      >
        Cancel
      </button>
      <button
        class="mocked-class"
        data-testid="common-button"
      >
        Confirm
      </button>
    </div>
  </div>
</DocumentFragment>
`;
